Changed attack evaluation order

kleiner-fisch opened the pull request:
This fixes #4041 

### Problem

There, it was reported that in one turn a Reaper attacked, dragging the player back, while dragging the player back, stepped on a shaft trap, which caused the Reaper to be shafted. 

This resulted in a message `DEAD MONSTER burns you.` because the brand damage was evaluated after the monster was shafted.

### Changes

Moved brand damage calculation from phase_hit to phase_damaged. The reason for this is to have `apply_damage_brand()` called right after `announce_hit()`, before `mons_attack_effects()` is called. The reason is that both methods are closely linked to the weapon, while `mons_attack_effects()` is not.

This reordering fixes #4041 because the disappearance of the attacker was caused by the dragging ability, which is handled in `mons_attack_effects()`.

The bug could also have been fixed by simply checking inside `apply_damage_brand()`, whether the attacker is still alive. However, that route seemed less sensible to me.

I also moved the call to `_apply_flux_contam()` because in its nature it seemed quite similar to a weapon brand.

kleiner-fisch committed:
Added comments

kleiner-fisch committed:
Changed when brand_damage is done during attacking

Moved brand damage calculation from phase_hit to phase_damaged.
The reason for this is to have `apply_damage_brand()` called right
after `announce_hit()`, before `mons_attack_effects()` is called. The
reason is that both methods are closely linked to the weapon, while
`mons_attack_effects()` is not.

This reordering has the benefit that it fixes #4041, where
`apply_damage_brand()` was called when the attacking monster
was shafted in the same turn it attacked.

The bug could also have been fixed by simply checking inside
`apply_damage_brand()`, whether the attacker is still alive. However,
that route seemed less sensible to me.

I also moved the call to `_apply_flux_contam()` because in its nature
it seemed quite similar to a weapon brand.

randomdevpete reviewed (commented):
I'm concerned about the big mechanical change of brands no longer getting to apply if damage is zero.

Apart from that, I've made a few comments where things can be tidied and clarified a bit.

I wonder if things could be further simplified by standardising *all* these functions to "return true if the attack should be aborted and/or the defender has died" (so added `return !defender->alive()` into more places, and also checking the result from more of these functions so we can avoid having ->alive() checks all over the place generally)

kleiner-fisch commented:
I appreciate your feedback!
Tbh, I moved `apply_damage_brand()` to get the output I thought sensible, without really understanding the consequences.

I will incorporate your feedback.

kleiner-fisch committed:
During attack store the attacker description

During an attack the attacker might suddenly disappear (e.g. be shafted
during an attack as in #4041).

To still have the description of the attacker available (e.g. to print attack
related messages after the attacker disappeared) we store the attacker
description during attack initialization.

kleiner-fisch committed:
Incorporated feedback from #4085

Main change is to move call to apply_damage_brand() back to
handle_phase_hit() to ensure brand damage is applied even when a hit
does not make damage

kleiner-fisch committed:
Unified return value of apply_damage_brand()

Now apply_damage_brand() returns true when the defender is banished,
which is consistent with the other usages of apply_damage_brand()

kleiner-fisch commented on crawl-ref/source/ranged-attack.cc:
```diff
@@ -401,6 +376,37 @@ int ranged_attack::calc_mon_to_hit_base()
+    {
+        return true;
+    }
+    if (using_weapon() && testbits(weapon->flags, ISFLAG_CHAOTIC)
```
I can only think of doing this by moving the code from `attack::apply_damage_brand` to a new method. Then `attack::apply_damage_brand` would do the chaos stuff, and call the new method.

But does that really improve the code?

kleiner-fisch committed:
linting: added braces

kleiner-fisch commented:
Incorporated the feedback.

Most importantly, brand damage is now again applied even when damage is 0.

### Implementation
To fix the original issue I added a class field that stores the attacker description. Then, even when the attacker disappears we can provide proper messages as to who does something.

CrawlOdds commented:
The underlying issue here has been fixed by not resetting monsters until the end of the turn, so closing this one.
